### ISSUE TITLE
ROP shortcuts

### ISSUE THREAD

**Issue by 152334H** [Gallopsled/pwntools#1681]
# tl;dr 
* `ROP.system_call()` for quick syscalls *without* using massive SIGROP
* `ROP.pop()` for quickly popping a few gadgets
* `ROP.pop.<reg>()` to quickly pop a single register
* All of these are things I have [personally implemented](https://github.com/152334H/pwnscripts/blob/dev/pwnscripts/rop.py) and tested (somewhat); this is to ask if these are reasonable features to add to pwntools itself
* Hopefully these shortcuts don't already exist in pwntools because that would be rather embarrassing

---

## Small Feature Request

CTFs often come with input limitations. Although it may be possible to do stack pivoting + SROP, it would be a lot simpler if there was a syscall method that *didn't* involve a massive 100 byte payload.

Proposition: a function to do exactly this

```python
def system_call(identifier, args):
    # add syscall `identifier` with arguments `args` to the ropchain
    # `identifier` can be a syscall number OR a "SYS_*" string resolvable by `pwnlib.constants`
```
Example output:
```python
>>> r = ROP(context.binary)
>>> r.system_call('read', [0, context.binary.bss(), 50])
>>> print(r.dump())
0x0000:         0x44a309 pop rdx; pop rsi; ret
0x0008:             0x32 [arg3] rdx = 50
0x0010:         0x4b92e0 [arg2] rsi = 4952800
0x0018:         0x41e4af pop rax; ret
0x0020:              0x0 [arg0] rax = SYS_read
0x0028:         0x401696 pop rdi; ret
0x0030:              0x0 [arg1] rdi = 0
0x0038:         0x4022b4 SYS_read
```
Compared with sigreturn:
```python
>>> r.call('read', [1,2,3])
[*] Using sigreturn for 'SYS_read'
>>> print(r.dump())
0x0000:         0x41e4af pop rax; ret
...<long output omitted>...
0x0108:              0x0 sigmask
```
---
Another smaller issue is that `ROP.setRegisters`... doesn't actually set registers? It might also be nice to have a `ROP.pop` method to quickly add a `pop` gadget to `ROP._chain`. 

```python
def pop(registers):
    # same input as setRegisters, but actually adds the gadgets to `ROP._chain`
def pop.<register>(value):
    # set a single register to `value` in `ROP._chain`
```
Examples [here](https://github.com/152334H/pwnscripts/blob/dev/pwnscripts/rop.py#L22-L30) and [here](https://github.com/152334H/pwnscripts/blob/dev/pwnscripts/rop.py#L35-L39)

## Implementation
As in the TL;DR, I already implemented this (among other helper functions) in my own personal [repo](https://github.com/152334H/pwnscripts/). For brevity, the pseudocode for the ROP modifications looks something like:

```python
class ROP(rop.rop.ROP):
    def pop(self, regs):
        # add regs to self.rop._chain
    def pop.__getattr__(self, reg, v): # This is implemented as a class instance in reality, because functions can't have .__getattr__()
        # run pop(v) on a single register `reg`
    def system_call(self, identifier, args):
        syscall = # find syscall gadget
        label = # use pwnlib.constants to resolve the syscall name from `identifier`
        # add syscall 'label' with `args` to rop chain
```
Granted, the code linked in the TL;DR can't be integrated immediately. It's written for python3-only and it's still lacking strong testing. I can try to make a PR with python2/3 code if these features sound OK.

**Comment by Arusekk** [comment]
Wow, I love those things! I missed them for long enough to appreciate this now. I will take a look at the code in a while.

**Comment by zachriggle** [comment]
Looks really neat!! and similar to something I've wanted to do for a while.  You're right that `setRegisters` does not actually set registers, it returns a stack that is used internally.  Worse, you can't even pass it to `ROP.raw()`. 

Here's a few more suggestions for ways to make this even cooler:

## Make `setRegisters` work with `rop.raw`

Ideally, this should work:

```
>>> rop = ROP(ELF.from_assembly('pop ebx; pop eax; ret'))
>>> rop.raw(rop.setRegisters({'ebx': 32}))
>>> rop.dump()
```

You'll need to implement `__flat__` for the `Gadget` and `Padding` classes, and possibly a little bit more work.

## ROP.register Callable

`ROP.register` currently returns a `Gadget` object.

```
>>> rop = ROP(ELF.from_assembly('pop ebx; pop eax; ret'))
>>> rop.eax
Gadget(0x10000000, ['pop eax', 'ret'], ['eax'], 0x8)
```

It would be neat if `Gadget` objects were a callable (implemented `__call__`), so you could do:

```
>>> rop.eax(0xdeadbeef)
```

In order to set the value of register `eax`.  I recommend using `ROP.setRegisters` to get the appropriate stack and then feeding it to `ROP.raw` (since it handles `add esp` and other offsets).  Note in my example, we can set `ebx` but it's not a simple `pop` so `setRegisters` should be used.

```
>>> rop.setRegisters({'ebx': 32})
[(268435456,
  Gadget(0x10000000, ['pop ebx', 'pop eax', 'ret'], ['ebx', 'eax'], 0xc)),
 (32, 'ebx'),
 (<pwnlib.rop.rop.Padding at 0x10b5612e0>, 'eax')]
```


## Support for more argument types

You stated you want to support both `'read'` and `SYS_read`, but I'd also recommend supporting `'SYS_read'` as a string.

I see you already added support such that `ROP.dump()` displays `SYS_read` in your example, which makes me super happy!

## Delegate to libc `syscall`

If `syscall` is a defined function (e.g. if we loaded `ROP()` with a libc), `system_call` should delegate to it where possible.

We can do this most trivially for the case where the name of the syscall is passed in (e.g. `'read'` or `'SYS_read'`) and looking up the function, but we can also do it if the `Constant` is passed in, since the name is a property of the constant:

```
>>> constants.SYS_read.s
'SYS_read'
```

Delegating to libc if it's available saves us from having to set `eax` / `rax`, which isn't necessarily a popular gadget and would be more "automagic" to users.

## Rename `system_call` to `syscall`

In support of the above "Delegate to libc syscall" it might be useful to rename the Python method.  Ultimately, it would mean that:

```
ROP.syscall(constants.SYS_exit)
```

Magically does the "best" thing:
* Use `libc.so`'s `syscall` function (if available)
* Otherwise, look for actual `syscall; ret` gadgets and do it as you've explained in your main post

**Comment by 152334H** [comment]
Going to close this for now. Implementing some of these suggestions is going to take mind-boggling amount of changes that I don't have the skill to do easily.

**Comment by zachriggle** [comment]
I didn't mean to scare you away! Just your ideas gave me some more.  We'd love to merge whatever you feel comfortable contributing!
